src/lib.rs

```diff
@@ -157,7 +167,15 @@ where F: FnOnce(&mut [MaybeUninit<u8>]) -> T
 	unsafe {
 	    let slice = slice::from_raw_parts_mut(allocad_ptr as *mut MaybeUninit<u8>, size);
 	    let callback = ManuallyDrop::take(&mut callback);
-	    rval = MaybeUninit::new(panic::catch_unwind(AssertUnwindSafe(move || callback(slice))));
+
+        #[cfg(feature = "no_std")]
+	    {
+            rval = MaybeUninit::new(callback(slice));
```

Comment:
It is possible to implement a custom `catch_unwind` in a `no_std` environment however there are some drawbacks. 

- aborting panics behave the same but unwinding panics within the closure will turn into aborting panics after being caught. This is because with no std present you must call onto OS-specific libraries to produce unwinding panics making it extremely hard to resume the unwinding process. 

- The same ordeal goes for the clean up of the panic 
`extern "C" {
    fn __rust_panic_cleanup(payload: *mut u8) -> *mut (dyn Any + Send + 'static);
}` is a core part of  `catch_unwind` and is not present without the std.

- The implementation of the custom `catch_unwind` directly calls `core::intrinsics::r#try` which is unstable and requires the use of nightly

With the custom `catch_unwind` any unwinding panic should be caught but unwinding will not be continued and an aborting panic will be produced. This is irreverent for most cases with no_std but in the rare cases that a unwinding panic is produced it will provide a safety net. The only cases where I can see this happening are  
- a downstream crate that enables the `no_std` feature for this crate while using std itself. 
- a downstream crate directly calls upon OS-specific unwinding functionality. In those cases aborting panics should also be produced.

> Once inline assembly is a stable feature I would like to replace the FFI trampoline with a #[naked] assembly-implemented dynamic stack allocation/deallocation function trampoline (or remove the primary trampoline entirely and just use inline asm! within the main function that calls into the closure) in pure Rust.
However, currently, having an FFI trampoline is the only way to implement dynamic stack growth in Rust on stable. Perhaps limiting safe no_std support to nightly only and doing this now would allow a fix of this problem but it would be unfortunate to make the no_std feature depend on a nightly compiler.

To my knowledge using the `no_std` feature requires the use of `lang_items` and `lang_items` requires the use of nightly so making the no_std feature only available on nightly seems like a non-issue


> If we should install our own panic handler when feature no_std is enabled to prevent an unwind by a downstream crate's use of std's default panic handler? I don't know if this would prevent downstream crates from installing their own panic handler entirely, I think it would but perhaps it's possible to install a panic handler per-crate, or have multiple panic handlers, or set our own default panic handler that can be overridden by a downstream crate that is #![no_std] itself (i.e. just prevent the std panic handler from being registered by the bottom-level crate of the dependency graph)? That's... probably not the case unfortunately but again I don't really know much about it...

You can only define a single panic handler per compiled program. That means if we add our own panic_handler it would prevent upstream crate from defining their own

Reply:
> Once inline assembly is a stable feature 

Inline assembly was stabilized in 1.59.0 for the following architectures
- x86 and x86-64
- ARM
- AArch64
- RISC-V

Reply:
Oh, thank you. There was a lot I didn't know there.
Huh, I did not know even know inline assembly was already stable...